cfgmgr/tunnelmgr.cpp

```diff
@@ -66,22 +169,160 @@ void TunnelMgr::doTask(Consumer &consumer)
     }
 }
 
-bool TunnelMgr::doIpInIpTunnelTask(const KeyOpFieldsValuesTuple & t)
+bool TunnelMgr::doTunnelTask(const KeyOpFieldsValuesTuple & t)
 {
     SWSS_LOG_ENTER();
 
-    const std::string & TunnelName = kfvKey(t);
+    const std::string & tunnelName = kfvKey(t);
     const std::string & op = kfvOp(t);
+    TunnelInfo tunInfo;
+
+    for (auto idx : kfvFieldsValues(t))
+    {
+        const std::string & field = fvField(idx);
+        const std::string & value = fvValue(idx);
+        if (field == "dst_ip")
+        {
+            tunInfo.dst_ip = value;
+        }
+        else if (field == "tunnel_type")
+        {
+            tunInfo.type = value;
+        }
+    }
 
     if (op == SET_COMMAND)
     {
-        m_appIpInIpTunnelTable.set(TunnelName, kfvFieldsValues(t));
+        if (tunInfo.type == IPINIP)
+        {
+            tunInfo.remote_ip = m_peerIp;
+
+            if (!m_peerIp.empty() && !configIpTunnel(tunInfo))
+            {
+                return false;
+            }
+            else if (m_peerIp.empty())
+            {
+                SWSS_LOG_NOTICE("Peer/Remote IP not configured");
+            }
+
+            m_appIpInIpTunnelTable.set(tunnelName, kfvFieldsValues(t));
+        }
+        m_tunnelCache[tunnelName] = tunInfo;
     }
     else
     {
-        m_appIpInIpTunnelTable.del(TunnelName);
+        auto it = m_tunnelCache.find(tunnelName);
+
+        if (it == m_tunnelCache.end())
+        {
+            SWSS_LOG_NOTICE("Tunnel %s not found", tunnelName.c_str());
+            return true;
+        }
+
+        tunInfo = it->second;
+        if (tunInfo.type == IPINIP)
+        {
+            m_appIpInIpTunnelTable.del(tunnelName);
+        }
+        m_tunnelCache.erase(tunnelName);
     }
     
-    SWSS_LOG_NOTICE("Tunnel %s task, op %s", TunnelName.c_str(), op.c_str());
+    SWSS_LOG_NOTICE("Tunnel %s task, op %s", tunnelName.c_str(), op.c_str());
+    return true;
+}
+
+bool TunnelMgr::doLpbkIntfTask(const KeyOpFieldsValuesTuple & t)
+{
+    SWSS_LOG_ENTER();
+
+    vector<string> keys = tokenize(kfvKey(t), config_db_key_delimiter);
+
+    /* Skip entry with just interface name. Need to handle only IP prefix*/
+    if (keys.size() == 1)
+    {
+        return true;
+    }
+
+    string alias(keys[0]);
+    IpPrefix ipPrefix(keys[1]);
+
+    m_intfCache[alias] = ipPrefix;
+
+    if (alias == LOOPBACK_SRC && !m_tunnelCache.empty())
+    {
+        int ret = 0;
+        std::string res;
+        ret = cmdIpTunnelIfAddress(ipPrefix.to_string(), res);
+        if (ret != 0)
+        {
+            SWSS_LOG_WARN("Failed to assign IP addr for tun if %s", ipPrefix.to_string().c_str());
```

Comment:
maybe printout of `res` would help with debugging

Reply:
Addressed